### ISSUE TITLE
NSData is not being externalized correctly

### ISSUE THREAD

**Issue by catshow** [nesium/cocoa-amf#2]
Per the spec at the bottom of page 11.  http://opensource.adobe.com/wiki/download/attachments/1114283/amf3_spec_05_05_08.pdf

It should be encoding the BinaryArray type marker 0xC then U29 with the length encoded then left shifted by 1.  What is really happening is it is encoding 0x4 (int) with 0xC (byte array marker), and then 0x4 (int) and then U29 length <<1.  for my test case i was encoding a 32 byte NSData object.

it ended up looking like this 0x04 0x0C 0x04 0x41 (... data ...) 

what it should be is 0x0C 0x41 (... data ...)

Thanks

**Comment by catshow** [comment]
I changed this in AMFArchiver.m
- (void)_encodeASObject:(ASObject *)value
  {
  [self encodeUnsignedChar:kAMF3ObjectType];
  if ([m_objectTable indexOfObjectIdenticalTo:value] != NSNotFound)
  {
      [self encodeUnsignedInt29:([m_objectTable indexOfObject:value] << 1)];
      return;
  }
  [m_objectTable addObject:value];
  AMF3TraitsInfo *traits = [[[AMF3TraitsInfo alloc] init] autorelease];
  traits.externalizable = value.isExternalizable;
  traits.dynamic = (value.type == nil || [value.type length] == 0);
  traits.count = (traits.dynamic || traits.externalizable ? 0 : [value count]);
  traits.className = value.type;
  traits.properties = (traits.dynamic ? nil : (id)[value.properties allKeys]);
  [self _encodeTraits:traits];
  
  if (value.isExternalizable)
  {
      for (id obj in value.data)
      {
          [self encodeObject:obj];
      }
  }
  
  for (NSString *key in value.properties)
  {
      if (traits.dynamic)
      {
          if (![key isKindOfClass:[NSString class]])
              key = [key description];
          [self _encodeString:key omitType:YES];
      }

==>     NSObject *o = [value.properties objectForKey:key];
        if ([o isKindOfClass:[NSData class]])
        {
            [self encodeObject:[AMFPlainData plainDataWithData:(NSData *)o]];
        }
        else
        {
            [self encodeObject:o];
==>     }
    }
    if (traits.dynamic)
    {
        [self encodeUnsignedInt29:((0 << 1) | 1)];
    }
}

and this which was the last part of -(void) encodeObject:

```
else if ([value isKindOfClass:[AMFPlainData class]])
{
    if (m_currentObjectToSerialize != nil)
    {
        [m_currentObjectToSerialize addObject:value];
        [self _ensureIntegrityOfSerializedObject];
        return;
    }
```

==>     [self encodeUnsignedChar:kAMF3ByteArrayType];
        NSData *d = [(AMFPlainData *)value data];
        [self encodeUnsignedInt29:(([d length] << 1) | 1)];
==>     [self encodeDataObject:d];
    }
    else
    {
        [self _encodeCustomObject:value];
    }
}

and this where I removed the lines that encoded the type and size
- (void)_encodeData:(NSData *)value
  {
  if ([m_objectTable indexOfObjectIdenticalTo:value] != NSNotFound)
  {
      [self encodeUnsignedInt29:([m_objectTable indexOfObject:value] << 1)];
      return;
  }
  ==>
  LINES REMOVED HERE
  ==> [m_objectTable addObject:value];
  [self encodeDataObject:value];
  }

**Comment by catshow** [comment]
That didn't work so well.  Just send me a note and I will send you the corrected file.

G

**Comment by nesium** [comment]
- applied gwatkins' patch (thanks!)
- closed by 8c36967ccd2f28f58b068b88c8d01ce20d00d8af (which was also reported by gwatkins)
- added unit test for #2
